core/appmodule/handler.go

```diff
@@ -0,0 +1,53 @@
+package appmodule
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+
+	"github.com/cosmos/cosmos-sdk/depinject"
+)
+
+// Handler describes an ABCI app module handler. It can be injected into a
+// depinject container as a one-per-module type (in the pointer variant).
+type Handler struct {
+	// Services are the msg and query services for the module. Msg services
+	// must be annotated with the option cosmos.msg.v1.service = true.
+	Services []ServiceImpl
+
+	// BeginBlocker doesn't take or return any special arguments as this
+	// is the most stable across Tendermint versions and most common need
+	// for modules. Special parameters can be inspected and/or returned
+	// using custom hooks that the app will provide which may vary from
+	// one Tendermint release to another.
+	BeginBlocker func(context.Context) error
+
+	// EndBlocker doesn't take or return any special arguments as this
+	// is the most stable across Tendermint versions and most common need
+	// for modules. Special parameters can be inspected and/or returned
+	// using custom hooks that the app will provide which may vary from
+	// one Tendermint release to another.
+	EndBlocker func(context.Context) error
```

Comment:
this would need to return []abci.ValidatorUpdate{}

Reply:
No, only one module does that and it's a special case that would be handled separately. Generally they don't

Reply:
ah okay, I see wasn't clear how that would be handled.

Reply:
I haven't quite figured out the design for that part but I know most modules don't need it and it introduces and abci dependency. So it might be coupled to a specific version of runtime/BaseApp

Reply:
I'm not following...These are ABCI methods, why can they not depend on ABCI? Also, I think we should be returning validator updates here.

Reply:
something would go for InitGenesis. I think the goal should be to not rely on tendermint types within modules but its unclear how to accomplish this, unless we create an sdk wrapper around the type.

Reply:
I'm not aware that we have tendermint wrapper types. Either way the point is that only one module needs to return validator updates and it would be an error for two modules to do so. So it's just less confusing and fewer tm dependencies to just treat staking as a special type of end blocker

Reply:
Also in abci++ or future versions the flow could potentially be different with respect to abci types. All the modules that don't need to touch that stuff (which is most modules) don't break because of an abci change

Reply:
> Either way the point is that only one module needs to return validator updates and it would be an error for two modules to do so.

I don't think it's safe to make this assumption. It is an assumption after all. You never know, an app might want to have the ability to merge and have preference orders for updates.

So I still don't see why we're not returning ABCI types here.